staging/src/k8s.io/api/apps/v1/types.go

```diff
@@ -114,6 +114,40 @@ type RollingUpdateStatefulSetStrategy struct {
 	Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
 }
 
+// PersistentVolumeClaimDeletePolicyType is a string enumeration of the policies that will determine
+// when volumes from the VolumeClaimTemplates will be deleted when the controlling StatefulSet is
+// deleted or scaled down.
+type PersistentVolumeClaimDeletePolicyType string
+
+const (
+	// RetainPersistentVolumeClaimDeletePolicyType is the default
+	// PersistentVolumeClaimDeletePolicy and specifies that
+	// PersistentVolumeClaims associated with StatefulSet VolumeClaimTemplates
+	// will not be deleted.
+	RetainPersistentVolumeClaimDeletePolicyType PersistentVolumeClaimDeletePolicyType = "Retain"
+	// DeletePersistentVolumeClaimDeletePolicyType specifies that
+	// PersistentVolumeClaims associated with StatefulSet VolumeClaimTemplates
+	// will be deleted in the scenario specified in
+	// StatefulSetPersistentVolumeClaimPolicy.
+	DeletePersistentVolumeClaimDeletePolicyType PersistentVolumeClaimDeletePolicyType = "Delete"
+)
+
+// StatefulSetPersistentVolumeClaimPolicy describes the policy used for PVCs
+// created from the StatefulSet VolumeClaimTemplates.
+type StatefulSetPersistentVolumeClaimPolicy struct {
+	// OnSetDeletion specifies what happens to PVCs created from StatefulSet
+	// VolumeClaimTemplates when the StatefulSet is deleted. The default policy
+	// of `Retain` causes PVCs to not be affected by StatefulSet deletion. The
+	// `Delete` policy causes those PVCs to be deleted.
+	OnSetDeletion PersistentVolumeClaimDeletePolicyType `json:"onSetDeletion,omitempty" protobuf:"bytes,1,opt,name=onSetDeletion,casttype=PersistentVolumeClaimDeletePolicyType"`
+	// OnScaleDown specifies what happens to PVCs created from StatefulSet
+	// VolumeClaimTemplates when the StatefulSet is scaled down. The default
+	// policy of `Retain` causes PVCs to not be affected by a scaledown. The
+	// `Delete` policy causes the associated PVCs for condemned pods to be
+	// deleted.
+	OnScaleDown PersistentVolumeClaimDeletePolicyType `json:"onScaleDown,omitempty" protobuf:"bytes,2,opt,name=onScaleDown,casttype=PersistentVolumeClaimDeletePolicyType"`
```

Comment:
Why are these omitempty?

Reply:
It's okay to not specify one (or both) of them, in which case the default "retain" will be filled in.

I was under the impression this is what omitempty was for, but I confess I was 50% copy-pasting from other fields...

LMK what would be better.

Reply:
In general omitempty is for the optional field itself - this field actually isn't optional (the struct is) and a user specifying this in a patch or returning the struct in JSON would expect OnScaleDown/OnSetDeletion to always be returned (at a minimum).  So we'd remove omitempty from these, add //+optional - and make sure you default the empty value here.

However, let me go make sure that's reasonably consistent with other recent decisions before you do anything (will try to get that sorted very soon)